File: frontend-sdk/src/core/registry.ts
import type { CanvasNodeData, NodeDefinition } from './types';

const normalizeDefinition = <TData extends CanvasNodeData>(definition: NodeDefinition<TData>): NodeDefinition => {
  if (!definition || typeof definition !== 'object') throw new TypeError('Node definition must be an object.');
  if (typeof definition.type !== 'string' || !definition.type.trim()) throw new Error('Node definition type is required.');
  if (typeof definition.title !== 'string' || !definition.title.trim()) throw new Error(`Node definition "${definition.type}" requires a title.`);
  if (typeof definition.category !== 'string' || !definition.category.trim()) throw new Error(`Node definition "${definition.type}" requires a category.`);
  if (!Array.isArray(definition.inputs) || !Array.isArray(definition.outputs)) {
    throw new TypeError(`Node definition "${definition.type}" requires input and output arrays.`);
  }
  if (typeof definition.createData !== 'function') {
    throw new TypeError(`Node definition "${definition.type}" requires a createData function.`);
  }
  if (definition.validate !== undefined && typeof definition.validate !== 'function') {
    throw new TypeError(`Node definition "${definition.type}" validate must be a function.`);
  }
  if (definition.execute !== undefined && typeof definition.execute !== 'function') {
    throw new TypeError(`Node definition "${definition.type}" execute must be a function.`);
  }
  const clonePorts = (ports: NodeDefinition['inputs'], direction: string) => Object.freeze(ports.map(port => {
    if (!port || typeof port !== 'object'
      || typeof port.id !== 'string' || !port.id.trim()
      || typeof port.label !== 'string' || !port.label.trim()
      || typeof port.dataType !== 'string' || !port.dataType.trim()) {
      throw new Error(`Node definition "${definition.type}" has an invalid ${direction} port.`);
    }
    if (port.required !== undefined && typeof port.required !== 'boolean') throw new TypeError(`${direction} port required must be boolean.`);
    if (port.multiple !== undefined && typeof port.multiple !== 'boolean') throw new TypeError(`${direction} port multiple must be boolean.`);
    return Object.freeze({ ...port });
  }));
  const inputs = clonePorts(definition.inputs, 'input');
  const outputs = clonePorts(definition.outputs, 'output');
  for (const [direction, ports] of [['input', inputs], ['output', outputs]] as const) {
    const ids = new Set<string>();
    for (const port of ports) {
      if (ids.has(port.id)) throw new Error(`Node definition "${definition.type}" has duplicate ${direction} port id "${port.id}".`);
      ids.add(port.id);
    }
  }
  return Object.freeze({
    ...definition,
    inputs,
    outputs,
  }) as unknown as NodeDefinition;
};

export class NodeRegistry {
  private readonly definitions = new Map<string, NodeDefinition>();
  private _revision = 0;

  get revision(): number {
    return this._revision;
  }

  register<TData extends CanvasNodeData>(definition: NodeDefinition<TData>): () => void {
    const normalized = normalizeDefinition(definition);
    if (this.definitions.has(normalized.type)) {
      throw new Error(`Node type "${normalized.type}" is already registered.`);
    }
    this.definitions.set(normalized.type, normalized);
    this._revision += 1;
    return () => {
      if (this.definitions.get(normalized.type) === normalized) this.unregister(normalized.type);
    };
  }

  replace<TData extends CanvasNodeData>(definition: NodeDefinition<TData>): void {
    const normalized = normalizeDefinition(definition);
    this.definitions.set(normalized.type, normalized);
    this._revision += 1;
  }

  unregister(type: string): boolean {
    const removed = this.definitions.delete(type);
    if (removed) this._revision += 1;
    return removed;
  }

  get(type: string): NodeDefinition | undefined {
    return this.definitions.get(type);
  }

  require(type: string): NodeDefinition {
    const definition = this.get(type);
    if (!definition) throw new Error(`Unknown node type "${type}".`);
    return definition;
  }

  list(): NodeDefinition[] {
    return [...this.definitions.values()];
  }

  has(type: string): boolean {
    return this.definitions.has(type);
  }
}

File: backend-sdk/schema/registry.go
package schema

import (
	"fmt"
	"sort"
	"strings"
	"sync"

	"github.com/flowcanvas/flowcanvas-backend-sdk/graph"
)

type PortDataType string

const (
	DataAny   PortDataType = "any"
	DataText  PortDataType = "text"
	DataImage PortDataType = "image"
	DataVideo PortDataType = "video"
	DataAudio PortDataType = "audio"
	DataJSON  PortDataType = "json"
)

type PortDefinition struct {
	ID       string       `json:"id"`
	Label    string       `json:"label"`
	DataType PortDataType `json:"dataType"`
	Required bool         `json:"required,omitempty"`
	Multiple bool         `json:"multiple,omitempty"`
}

type NodeValidator func(node graph.Node) []graph.ValidationIssue

type NodeDefinition struct {
	Type               string           `json:"type"`
	Title              string           `json:"title"`
	Category           string           `json:"category"`
	Description        string           `json:"description,omitempty"`
	Color              string           `json:"color,omitempty"`
	Icon               string           `json:"icon,omitempty"`
	Inputs             []PortDefinition `json:"inputs"`
	Outputs            []PortDefinition `json:"outputs"`
	RequiredDataFields []string         `json:"requiredDataFields,omitempty"`
	Validate           NodeValidator    `json:"-"`
}

type Registry struct {
	mu          sync.RWMutex
	definitions map[string]NodeDefinition
}

func NewRegistry(definitions ...NodeDefinition) (*Registry, error) {
	r := &Registry{definitions: map[string]NodeDefinition{}}
	for _, definition := range definitions {
		if err := r.Register(definition); err != nil {
			return nil, err
		}
	}
	return r, nil
}

func MustNewRegistry(definitions ...NodeDefinition) *Registry {
	r, err := NewRegistry(definitions...)
	if err != nil {
		panic(err)
	}
	return r
}

func (r *Registry) Register(definition NodeDefinition) error {
	if r == nil {
		return fmt.Errorf("registry is nil")
	}
	definition.Type = strings.TrimSpace(definition.Type)
	if definition.Type == "" {
		return fmt.Errorf("node definition type is required")
	}
	if definition.Title == "" {
		definition.Title = definition.Type
	}
	if definition.Inputs == nil {
		definition.Inputs = []PortDefinition{}
	}
	if definition.Outputs == nil {
		definition.Outputs = []PortDefinition{}
	}
	if err := validatePorts(definition.Inputs, "input", definition.Type); err != nil {
		return err
	}
	if err := validatePorts(definition.Outputs, "output", definition.Type); err != nil {
		return err
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.definitions == nil {
		r.definitions = map[string]NodeDefinition{}
	}
	if _, exists := r.definitions[definition.Type]; exists {
		return fmt.Errorf("node definition already registered: %s", definition.Type)
	}
	r.definitions[definition.Type] = definition
	return nil
}

func (r *Registry) MustRegister(definition NodeDefinition) {
	if err := r.Register(definition); err != nil {
		panic(err)
	}
}

func (r *Registry) Get(nodeType string) (NodeDefinition, bool) {
	if r == nil {
		return NodeDefinition{}, false
	}
	r.mu.RLock()
	defer r.mu.RUnlock()
	definition, ok := r.definitions[nodeType]
	return definition, ok
}

func (r *Registry) Definitions() []NodeDefinition {
	if r == nil {
		return nil
	}
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]NodeDefinition, 0, len(r.definitions))
	for _, definition := range r.definitions {
		out = append(out, definition)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Type < out[j].Type })
	return out
}

func Compatible(source, target PortDataType) bool {
	return source == DataAny || target == DataAny || source == target
}

func (d NodeDefinition) Input(id string) (PortDefinition, bool) {
	for _, port := range d.Inputs {
		if port.ID == id {
			return port, true
		}
	}
	return PortDefinition{}, false
}

func (d NodeDefinition) Output(id string) (PortDefinition, bool) {
	for _, port := range d.Outputs {
		if port.ID == id {
			return port, true
		}
	}
	return PortDefinition{}, false
}

func validatePorts(ports []PortDefinition, kind, nodeType string) error {
	seen := map[string]struct{}{}
	for _, port := range ports {
		port.ID = strings.TrimSpace(port.ID)
		if port.ID == "" {
			return fmt.Errorf("%s port id is required for node type %s", kind, nodeType)
		}
		if port.DataType == "" {
			return fmt.Errorf("%s port %s dataType is required for node type %s", kind, port.ID, nodeType)
		}
		if _, exists := seen[port.ID]; exists {
			return fmt.Errorf("duplicate %s port %s for node type %s", kind, port.ID, nodeType)
		}
		seen[port.ID] = struct{}{}
	}
	return nil
}
